docs/design/variadics.md

```diff
@@ -523,113 +450,389 @@ expansion itself is relatively straightforward:
--   = `(<T, 1:Exactly(1)>, <T, |T|-1:AtLeast(0)>)` (splitting rule) (Note that
-    in this example we don't need to rewrite the representative expressions in
-    the last step, because the representative doesn't depend on `$I`).
-
-In general, we can split out up to `N` leading or trailing singular segments
-from a variadic segment that has arity type `AtLeast(N)`. So if the pattern has
-`N` leading singular segments, and the scrutinee has `M` leading singular
-segments where `M < N`, step 2 verifies that the first variadic scrutinee
-segment's arity type is a subtype of `AtLeast(N-M)`, and splits out `N-M`
-leading singular segments from it. It then applies the same procedure to last
-variadic scrutinee segment to ensure there are enough trailing singular
-segments.
-
-The rewrites in step 2 do not take place when pattern matching as part of a
-function call: we assume that physically separate function parameters are
-logically separate as well, so if they are not separate at the callsite, that
-likely indicates a bug.
-
-#### Step 3: Pattern adjustment
-
-Step 2 ensures that at the start of step 3 we have a subsequence of scrutinee
-segments which contain all the elements that the variadic pattern segment will
-match, and which cannot match any other pattern segment. Furthermore, the arity
-of the pattern segment that we deduced in step 1 must be equal to the sum of the
-arities of that scrutinee subsequence. As a result, step 3 can apply the
-splitting rule to decompose the pattern segment into a sequence with the same
-shape as the scrutinee subsequence.
-
-Note that any pack indexing expression in the original variadic pattern segment
-is guaranteed to have the form `B[:$I:]` (with no offset), where `B` is a
-deduced pack binding. The decomposition in this step will rewrite that
-expression in each segment to `B[:$I + A:]` (or `B[:A:]` if the segment is
-singular), where `A` is the sum of the arities of the earlier segments in the
-decomposition. This has the effect of decomposing the unknown contents of `B`
-into a sequence of segments with unknown representatives that has the same shape
-as the scrutinee subsequence.
-
-#### Step 4: Representative deduction
-
-Since the pattern and scrutinee now have identical shapes, each pattern segment
-is guaranteed to match all of the elements of the corresponding scrutinee
-segment. Consequently, we can perform type deduction segment-wise, deducing the
-representative type of each pattern segment from the representative type of the
-corresponding scrutinee segment.
-
-Within a segment, type deduction takes place as normal. The only aspect that's
-unique to variadics is that when unifying a pattern type expression
-`B[:$I + A:]` or `B[:A:]` with a scrutinee expression `S`, we deduce that `S` is
-the representative of the corresponding segment of `B`.
+The function type derived from that signature is:
+
+```carbon
+(Vector(First), Vector(Second), ... Vector(each Next))
```

Comment:
There are two kinds of functions: primitive function calls and families of function calls. A primitive function call has concrete parameter arity and types and no deduced parameters. Since you are concerned with variadic functions, primitive function calls are not enough, you need to be concerned about the families of function calls you get by having deduced parameters.

I grappled with this issue in [the "Compile-time function parameters" section of "Carbon: function overloading"](https://docs.google.com/document/d/1rTsDJ_ZNENH0mhi5E64MGjKJ29OYiI_dzBRz2GzYySU/edit?resourcekey=0-z6_xQ9xuJ5CZ6qJqiuSxuw&tab=t.0#heading=h.kx0d18wr7ba0). Variadics are trying to do the same thing: describe a family of implementations of the `Call` operator for a particular type, so:

```
fn F[... each T:! type](... each x: each T);
```

becomes:

```
class __typeof_F {}
let F:! __typeof_F = {}; 
final impl forall [... each T:! type] __typeof_F as Call(... each x: each T);
```

and then a particular call to `F` does "overload resolution" by looking to see if a matching implementation of `Call` exists for `__typeof_F`:

```
let a: i32 = 1;
let b: bool = true;
F(a, b);
```

is translated to:

```
let a: i32 = 1;
let b: bool = true;
F.((__typeof_F as Call(i32, bool)).Op)(a, b);
```

which finds the implementation of `Call` for `__typeof_F` with `each T` equal to `i32`, `bool`.

So the type of `F` is straightforward (`__typeof_F`), but not interesting (it is an empty class). It encapsulates all the overloads, but in a trivial way, since all values of that type are the same. What you are probably talking about, when you think of "function type", is some requirement that this type satisfies that says it is callable with some argument types, as in:

```
fn HasFnArg[T:! Call(i32, bool)](f: T) {
  let a: i32 = 1;
  let b: bool = true;
  f(a, b);
}

HasFnArg(F);
```

However, there is no way to say `T` implements a family of implementations of `Call` -- which might be written like `fn HasGenericFnArg[T:! forall [U:! type] Call(U, U) where .Result = Vector(U)](f: T)` -- and that is not expected to change soon. One issue is that it is hard in general to prove that a particular type would satisfy this condition -- for example, there could be a specialization somewhere that implements `Call` for that type with a different return type (and it wouldn't even necessarily be visible at a call site to `HasGenericFnArg`). Functions are the rare case where we might be able to do it, since behind the scenes they define a `final` implementation of `Call` (which corresponds to only supporting "closed" overloading by default). The other big issue is even with a `final` implementation, it would still be quite difficult to validate that a particular `forall` condition is satisfied, which violates the principle that the compiler shouldn't be asked to solve hard problems.

So, I would avoid talking about "function types" as a thing that captures how a function can be called. Instead, just talk about the signature of a particular function declaration.

Reply:
> So, I would avoid talking about "function types" as a thing that captures how a function can be called. Instead, just talk about the signature of a particular function declaration.

OK, done. That turned out to be less messy than I feared, and it even lets me get rid of "synthetic deduced parameters".